Enhance junit log to include customized individual assertion rules

waynechen55 opened the pull request:
1. Introduce $overall_status to indicate the overall pass/fail status

2. In analyzeResult(), initialize all testcase status counters to zero, then count up respective status counter by iterating over all testcases executed on all vm guests, at last handle corner situation in which test run did not fail at predefined step. By the way, the undefined command/step at which the test run failed is tored in /root/commands_history file obtained from "history -w". In order to keep clean and shorten the history file, "history -c" is called before executing every modules.
    
3. Extract junit log generation functionality for individual testcase into a separate subroutine generate_testcase(), which also handles legacy test scenairos like guest_installation_run.
    
4. Output vm guest name in failure element in testcase xml if any failure.
    
5. Some unused elements in JUnit log are given "n/a".
    
6. Correct ambiguous usage of "time" element in all data structures. Because "time" is keyword in Perl, avoid use it and use "test_time" instead. And improve all time output formats in JUnit log, which improves the legacy test scenarios junit log output like guest_installation_run so both total time spent and time cost for individual vm guest are recorded. Time cost display for newly developed feature test is also supported, although the author can decide whether use it or not.
    
7. The system error element also support newly developed feature test. And the author can decide whether use it or not. The same also applies to system out element.
    
8. Call junit_log_provision before any following operations in post_fail_hook to ensure JUnit log is alwasy available under any circumstances.
    
9. Please set assertion rules to be used with this any newly developed feature test like $self->{test_results}->{$guest}->{TEST1}->{status} = 'SKIPPED/FAILED/PASSED/SOFTFAILED/TIMEOUT/UNKNOWN'. Time spent for each test and corresponding error can also be given by using $self->{test_results}->{$guest}->{TEST1}->{test_time} = 'XXmYYs', $self->{test_results}->{$guest}->{TEST1}->{error} can be given "SPECIFIC SYSTEM ERROR" and $self->{test_results}->{$guest}->{TEST1}->{output} can be given "SPECIFIC SYSTEM OUTPUT" which are not mandatory and can be omitted in feature test design.

- Related ticket: n/a
- Needles: https: n/a
- Verification run: 
    1. sles15sp2 Run1 http://10.67.19.99/tests/700
    2.  sles15sp2 Run2 http://10.67.19.99/tests/701
    3.  sles11sp4 Run http://10.67.19.99/tests/649
    4.  Migration Run http://10.67.19.99/tests/702

alice-suse commented:
@waynechen55 Thanks for the great work! Some questions below:

- For the failure of sub-tests, is it planned to be shown in junit? Like original prj3 junit log, it can show the errors as  system-err in
https://openqa.nue.suse.com/tests/3716305#step/1_Guest_Migration_Test/5. In your verification jobs, eg http://10.67.19.99/tests/652#step/virsh_internal_snapshot/19, this key error msg is not shown in http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2.

- In this PR, I can see you give a lot of effort on test time, but it is not shown in the final junit. Why? IMO, if you do it, then it is better to show it in junit, like http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2, since it can be helpful. But if you do not want to show it, then better save the effort in code also.

- This PR also changes code in other test files like guest migration. Suggest to do verification run for all affected test projects(at least unit test for junit part). If no available workers, can ask @Julie-CAO  for help to retrigger on osd for final check.

alice-suse commented:
@Julie-CAO @guoxuguang Please also help to make a review. As we discussed during meeting for this topic, you must already be quite clear about the purpose of this PR. Please have a look whether it meets the requirement, if it introduces any concerns to our legacy test projects like prj1-prj6, and whether you are clear about how to give the junit log in the newly implemented virtualization feature tests like virtual network management , snapshot, etc.

Julie-CAO commented:
As for as I know, guest_migration_source_migrate.pm is not in use for prj3 tests, though I don't know the reason. guest_migration_src/dst are used instead.

Julie-CAO commented:
I assume the tests in this PR just set the examples how the junit report looks like. When we add tests in the future, we need set the $self->{test_results}->{$guest}{TEST1} = 'SKIPPED/FAILED/PASSED/SOFTFAILED/TIMEOUT/UNKNOWN' based on some core test commands in test scenarios. Is it correct? Do we need pass more info to junit report, such as error msg? Are there any examples? 
And do we need do any edits to existing tests to work with this PR?

waynechen55 commented:
Description and verification run updated. Please have a look at the latest situation.

waynechen55 commented:
> @waynechen55 Thanks for the great work! Some questions below:
> 
>     * For the failure of sub-tests, is it planned to be shown in junit? Like original prj3 junit log, it can show the errors as  system-err in
>       https://openqa.nue.suse.com/tests/3716305#step/1_Guest_Migration_Test/5. In your verification jobs, eg http://10.67.19.99/tests/652#step/virsh_internal_snapshot/19, this key error msg is not shown in http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2.
> 
>     * In this PR, I can see you give a lot of effort on test time, but it is not shown in the final junit. Why? IMO, if you do it, then it is better to show it in junit, like http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2, since it can be helpful. But if you do not want to show it, then better save the effort in code also.
> 
>     * This PR also changes code in other test files like guest migration. Suggest to do verification run for all affected test projects(at least unit test for junit part). If no available workers, can ask @Julie-CAO  for help to retrigger on osd for final check.

1. Please refer to verification run I and II which contains error messages output. I extended data structure to support this.
2. The time thing is already displayed in verification run, especially in guest_installation part which outputs such information always, so we can take advantage of this to output more meaningful test info. And "time" is also an important and integral part of junit log itself, so it is better to give accurate time information. Additionally, in order to demonstrate this also works for newly developed feature test, I added extra verification run I and II to prove. And as author of any feature test, he/she can decide whether to use it or not for individual test, because this can be omitted. Please refer to the last item in description to get more info about this.
3. Actually I do not change any underlying design logic used by all legacy test scenarios, basically no verification run is needed. But in order to make things more convincing, I added verification run IV for migration which looks as good as before.

waynechen55 commented:
> I assume the tests in this PR just set the examples how the junit report looks like. When we add tests in the future, we need set the $self->{test_results}->{$guest}{TEST1} = 'SKIPPED/FAILED/PASSED/SOFTFAILED/TIMEOUT/UNKNOWN' based on some core test commands in test scenarios. Is it correct? Do we need pass more info to junit report, such as error msg? Are there any examples?
> And do we need do any edits to existing tests to work with this PR?

1. I only changed the element name to "test_time" in guest_migration_source_migrate.pm in order to sync with base module and ensure it can always be used as normal and as before. It is trivial and very straightforward.
2. Please refer to verification runs I, II and IV. You can pass error message now. As far as the usage is concerned, please refer to the last item in description.

alice-suse commented:
> > @waynechen55 Thanks for the great work! Some questions below:
> > ```
> > * For the failure of sub-tests, is it planned to be shown in junit? Like original prj3 junit log, it can show the errors as  system-err in
> >   https://openqa.nue.suse.com/tests/3716305#step/1_Guest_Migration_Test/5. In your verification jobs, eg http://10.67.19.99/tests/652#step/virsh_internal_snapshot/19, this key error msg is not shown in http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2.
> > 
> > * In this PR, I can see you give a lot of effort on test time, but it is not shown in the final junit. Why? IMO, if you do it, then it is better to show it in junit, like http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2, since it can be helpful. But if you do not want to show it, then better save the effort in code also.
> > 
> > * This PR also changes code in other test files like guest migration. Suggest to do verification run for all affected test projects(at least unit test for junit part). If no available workers, can ask @Julie-CAO  for help to retrigger on osd for final check.
> > ```
> 
> 1. Please refer to verification run I and II which contains error messages output. I extended data structure to support this.

Cool! I can see it in verification 2) job. 

For the below content in that job, I guess you plan to put the error as "system-err". How about "system-out" and "failure"? They also have contents, but seems not so reasonable content for these two category. Can you help explain here your plan for them? Keep on purpose or inherit from before? Can current design support filling correct contents?(Actually IMO, IF you do not plan to support them, and n/a can be given by feature tests from the api, it is also ok at current stage)

# virsh xxxxxx snapshot-revert sle15sp2 sle15sp2 sles-15-sp2-64-fv-def-net --snapshotname internal-snapshot-sle15sp2-01 ; echo VlaqH-$?- > /dev/sshserial
# system-err: 

error: unknown command: xxxxxx
# system-out: 

sle15sp2 sles-15-sp2-64-fv-def-net 
# failure: 

sle15sp2 sles-15-sp2-64-fv-def-net

> 2. The time thing is already displayed in verification run, especially in guest_installation part which outputs such information always, so we can take advantage of this to output more meaningful test info. And "time" is also an important and integral part of junit log itself, so it is better to give accurate time information. Additionally, in order to demonstrate this also works for newly developed feature test, I added extra verification run I and II to prove. And as author of any feature test, he/she can decide whether to use it or not for individual test, because this can be omitted. Please refer to the last item in description to get more info about this.

I checked the verification jobs 1 and 2, time is shown for guest installation one. But for virsh_internal_snapshot, I do not see it. I can understand you leave the right to test writer. But since it is useful information, why not  always show it? :)


> 3. Actually I do not change any underlying design logic used by all legacy test scenarios, basically no verification run is needed. But in order to make things more convincing, I added verification run IV for migration which looks as good as before.
Great! Thanks!

waynechen55 commented:
> > > @waynechen55 Thanks for the great work! Some questions below:
> > > ```
> > > * For the failure of sub-tests, is it planned to be shown in junit? Like original prj3 junit log, it can show the errors as  system-err in
> > >   https://openqa.nue.suse.com/tests/3716305#step/1_Guest_Migration_Test/5. In your verification jobs, eg http://10.67.19.99/tests/652#step/virsh_internal_snapshot/19, this key error msg is not shown in http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2.
> > > 
> > > * In this PR, I can see you give a lot of effort on test time, but it is not shown in the final junit. Why? IMO, if you do it, then it is better to show it in junit, like http://10.67.19.99/tests/652#step/1_virsh_internal_snapshot/2, since it can be helpful. But if you do not want to show it, then better save the effort in code also.
> > > 
> > > * This PR also changes code in other test files like guest migration. Suggest to do verification run for all affected test projects(at least unit test for junit part). If no available workers, can ask @Julie-CAO  for help to retrigger on osd for final check.
> > > ```
> > 
> > 
> > 
> > 1. Please refer to verification run I and II which contains error messages output. I extended data structure to support this.
> 
> Cool! I can see it in verification 2) job.
> 
> For the below content in that job, I guess you plan to put the error as "system-err". How about "system-out" and "failure"? They also have contents, but seems not so reasonable content for these two category. Can you help explain here your plan for them? Keep on purpose or inherit from before? Can current design support filling correct contents?(Actually IMO, IF you do not plan to support them, and n/a can be given by feature tests from the api, it is also ok at current stage)
> 
> ```
> # virsh xxxxxx snapshot-revert sle15sp2 sle15sp2 sles-15-sp2-64-fv-def-net --snapshotname internal-snapshot-sle15sp2-01 ; echo VlaqH-$?- > /dev/sshserial
> # system-err: 
> 
> error: unknown command: xxxxxx
> # system-out: 
> 
> sle15sp2 sles-15-sp2-64-fv-def-net 
> # failure: 
> 
> sle15sp2 sles-15-sp2-64-fv-def-net
> ```
> 
> > 1. The time thing is already displayed in verification run, especially in guest_installation part which outputs such information always, so we can take advantage of this to output more meaningful test info. And "time" is also an important and integral part of junit log itself, so it is better to give accurate time information. Additionally, in order to demonstrate this also works for newly developed feature test, I added extra verification run I and II to prove. And as author of any feature test, he/she can decide whether to use it or not for individual test, because this can be omitted. Please refer to the last item in description to get more info about this.
> 
> I checked the verification jobs 1 and 2, time is shown for guest installation one. But for virsh_internal_snapshot, I do not see it. I can understand you leave the right to test writer. But since it is useful information, why not always show it? :)
> 
> > 1. Actually I do not change any underlying design logic used by all legacy test scenarios, basically no verification run is needed. But in order to make things more convincing, I added verification run IV for migration which looks as good as before.
> >    Great! Thanks!

1. system error is for error message and failure is for the affected subject, for example, vm guest. Currently no further work is needed on these.
2. Verification I for internal snapshot already shows time info for the second test item.

waynechen55 commented:
Updated description 7,9 and verification runs I, II and IV.  Suggest merge if no other big concerns.

guoxuguang commented:
LGTM

alice-suse commented:
LGTM.

@Julie-CAO @guoxuguang Please also have a look. If you both  have no other concerns, will merge it.

Julie-CAO commented:
LGTM
